packages/connect/src/base/stun.ts

```diff
@@ -82,109 +82,98 @@ type Address = {
-    verbose(
-      `Using own STUN servers did not lead to a result. Trying ${usableMultiaddrs
-        .map((ma: Multiaddr) => ma.toString())
-        .join(', ')}`
-    )
+  let responses: RequestWithResponse[] = await performSTUNRequests(
+    selectedStunServers,
+    socket,
+    STUN_TIMEOUT,
+    runningLocally
+  )
+
+  if (multiaddrs.length > DEFAULT_PARALLEL_STUN_CALLS) {
+    while (responses.length < 2) {
+      for (const selected of selectedStunServers) {
+        usedStunServers.add(selected.toString())
+      }
 
-    responses.push(...(await performSTUNRequests(usableMultiaddrs, socket, STUN_TIMEOUT)))
-  }
+      if (usedStunServers.size >= maxAttempts) {
+        break
+      }
 
-  if (responses.length == 0) {
-    // Even contacting public STUN servers did not lead
-    return
-  }
+      const toFetch = Math.min(maxAttempts, DEFAULT_PARALLEL_STUN_CALLS + usedStunServers.size) - usedStunServers.size
 
-  let filteredResults = getUsableResults(responses, runningLocally)
+      selectedStunServers = randomSubset(multiaddrs, toFetch, (ma: Multiaddr) => !usedStunServers.has(ma.toString()))
 
-  if ([0, 1].includes(filteredResults.length) && !runningLocally && !usingPublicServers) {
-    // Received results were not usable to determine public IP address
-    // now trying public ones
-    verbose(
-      `Using own STUN servers did not lead to a result. Trying ${usableMultiaddrs
-        .map((ma: Multiaddr) => ma.toString())
-        .join(', ')}`
-    )
-    usingPublicServers = true
-    usableMultiaddrs = randomSubset(PUBLIC_STUN_SERVERS, DEFAULT_PARALLEL_STUN_CALLS)
-    responses.push(...(await performSTUNRequests(usableMultiaddrs, socket, STUN_TIMEOUT)))
+      responses.push(...(await performSTUNRequests(selectedStunServers, socket, STUN_TIMEOUT, runningLocally)))
 
-    filteredResults = getUsableResults(responses, runningLocally)
+      if (selectedStunServers.length < toFetch) {
+        break
+      }
+    }
   }
 
-  const interpreted = intepreteResults(filteredResults)
-
-  if (interpreted.ambiguous) {
-    return undefined
-  } else {
-    return interpreted.publicAddress
-  }
+  return responses
 }
 
-async function performSTUNRequests(
+/**
+ * Performs STUN requests and returns their responses, if any
+ * @param multiAddrs STUN servers to contact
+ * @param socket socket to send requests and receive responses
+ * @param timeout STUN timeout
+ * @param runningLocally [optional] enable STUN local-mode
+ * @returns the responses, if any
+ */
+export async function performSTUNRequests(
   multiAddrs: Multiaddr[],
   socket: Socket,
-  timeout = STUN_TIMEOUT
-): Promise<Request[]> {
-  let requests = generateRequests(multiAddrs)
-  let results = decodeIncomingSTUNResponses(requests, socket, timeout)
+  timeout = STUN_TIMEOUT,
+  runningLocally = false
+): Promise<RequestWithResponse[]> {
+  const requests = generateRequests(multiAddrs)
+  const results = decodeIncomingSTUNResponses(requests, socket, timeout)
```

Comment:
Any particular reason we start the promises that decode the responses before we actually send the requests?
Honoring the causality in code can improve readability.

Reply:
You need to attach the event handler before you expect the event to come in.

Reply:
added a comment